### PR TITLE
Add unit test coverage for modules blocks.chain - Closes #1549

### LINKED ISSUE #1549: Add unit test coverage for modules blocks.chain
Parent #1301 

Chain methods:
- constructor
- saveGenesisBlock
- saveBlock
- __private.afterSave
- deleteBlock
- deleteAfterBlock
- applyGenesisBlock
- __private.applyTransaction
- applyBlock
- broadcastReducedBlock
- __private.popLastBlock
- deleteLastBlock
- recoverChain
- onBind

### PR DESCRIPTION
**Pull request by LucasIsasmendi**
### What was the problem?
Missed unit test for modules blocks chain.

### How did I fix it?
Moved current test to functional.
Added unit tests to skeleton file.

### Review checklist
* The PR solves #1549 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: test/functional/modules/blocks/chain.js [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+		timestamp,
+		previousBlock: blocksModule.lastBlock.get(),
+		transactions,
+	});
+	newBlock.id = blockLogic.getId(newBlock);
+	newBlock.height = previousBlock ? previousBlock.height + 1 : 1;
+	return newBlock;
+}
+
+describe('blocks/chain', () => {
```
Comment (karmacoma) [inline test/functional/modules/blocks/chain.js]: - Why test file is not located under `test/functional/system/blocks`?
- Why a test skeleton which describe unit tests? Seems like we only describe block application anyway.

**Comment by MaciejBaj** [comment]
Should be merged after https://github.com/LiskHQ/lisk/pull/1729 as bug fixing code changes are introduced over there.

File: test/unit/modules/blocks/chain.js [hunk 2]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+		});
 
-		it(
-			'should call modules.blocks.utils.getBlockProgressLogger with block.transactions.length / 100'
-		);
+		describe('when block.transactions is empty', () => {
+			it('should return resolved promise with no error', done => {
+				__private
+					.applyUnconfirmedStep(
+						{ id: 6, height: 6, transactions: [] },
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: we have 3-4 type of blocks. We should define these blocks and reuse them in these tests.

File: test/unit/modules/blocks/chain.js [hunk 3]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+						return expect(loggerStub.error.args[1][1]).to.deep.equal({
+							id: 1,
+							type: 0,
+							senderPublicKey: 'a',
+						});
+					});
+					it('should return rejected promise with error', done => {
+						__private
+							.applyUnconfirmedStep(
+								{
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: here as well.

File: test/unit/modules/blocks/chain.js [hunk 4]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+								return expect(loggerStub.error.args[1][1]).to.deep.equal({
+									id: 1,
+									type: 0,
+									senderPublicKey: 'a',
+								});
+							});
+							it('should return rejected promise with error', done => {
+								__private
+									.applyUnconfirmedStep(
+										{
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 5]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+								return expect(
+									modules.transactions.applyUnconfirmed.callCount
+								).to.equal(2);
+							});
 
-					it('should assign block');
+							it('should return resolved promise with no error', done => {
+								__private
+									.applyUnconfirmedStep(
+										{
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 6]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+		});
+	});
 
-				it('should call process.exit with 0');
+	describe('__private.applyConfirmedStep', () => {
+		describe('when block transaction is undefined', () => {
+			it('should return rejected promise with error', done => {
+				__private
+					.applyConfirmedStep(
+						{ id: 6, height: 6, transactions: undefined },
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 7]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+					});
 			});
+		});
 
-			describe('when modules.accounts.setAccountAndGet succeeds', () => {
-				it('should call __private.applyTransaction');
+		describe('when block transaction is empty', () => {
+			it('should return resolved promise with no error', done => {
+				__private
+					.applyConfirmedStep({ id: 6, height: 6, transactions: [] }, dbStub.tx)
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 8]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+							type: 0,
+							senderPublicKey: 'a',
+						});
+					});
 
-				it('should call __private.applyTransaction with sender');
+					it('should return rejected promise with error', done => {
+						__private
+							.applyConfirmedStep(
+								{
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 9]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+									senderPublicKey: 'a',
+								});
+							});
 
-				it('should call callback with error = undefined');
+							it('should return rejected promise with error', done => {
+								__private
+									.applyConfirmedStep(
+										{
+											id: 5,
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: here as well.

File: test/unit/modules/blocks/chain.js [hunk 10]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+									3,
+									'apply-ERR',
+									null
+								);
+							});
 
-			describe('after loop through block.transactions', () => {
-				it('should set genesis block as last block');
+							afterEach(() => {
+								expect(modules.accounts.getAccount.callCount).to.equal(1);
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: i don't like these assertions in afterEach. If we want to have something like this, we can create a function, and call it from all tests.

File: test/unit/modules/blocks/chain.js [hunk 11]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
-				it('should call callback with result = undefined');
+						describe('when succeeds', () => {
+							beforeEach(() => {
+								return modules.transactions.apply.callsArgWith(3, null, true);
+							});
+							it('should return resolved promise with no error', done => {
+								__private
+									.applyConfirmedStep(
+										{
+											id: 5,
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 12]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+			};
+			done();
 		});
 
 		afterEach(done => {
-			previousBlock = block;
+			blocksChainModule.saveBlock = saveBlockTemp;
+			expect(modules.blocks.lastBlock.set.calledOnce).to.be.true;
+			expect(modules.blocks.lastBlock.set.args[0][0]).to.deep.equal({
+				id: 5,
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 13]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+			library.db.tx = (desc, tx) => {
+				return tx();
+			};
+			done();
 		});
 
 		afterEach(done => {
-			previousBlock = block;
+			blocksChainModule.saveBlock = saveBlockTemp;
+			expect(modules.blocks.lastBlock.set.calledOnce).to.be.true;
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: assertions here as well

File: test/unit/modules/blocks/chain.js [hunk 14]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
-					expect(data).to.have.lengthOf(1);
-					expect(data[0].id).to.be.equal(block.id);
-					done(err);
+				it('should call a callback with error', done => {
+					__private
+						.saveBlockStep(
+							{
+								id: 5,
+								height: 5,
+								transactions: [{ id: 1, type: 0 }, { id: 2, type: 1 }],
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 15]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
-				}
-			};
-
-			var connect = sinonSandbox.stub();
-			var disconnect = sinonSandbox.stub();
-
-			db.$config.options.connect = connect;
-			db.$config.options.disconnect = disconnect;
+					expect(blocksChainModule.saveBlock.args[0][0]).to.deep.equal({
+						id: 5,
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 16]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+						beforeEach(() => {
+							return modules.rounds.tick.callsArgWith(1, null, true);
+						});
 
-				blocksModule.shared.getBlocks({ id: block.id }, err => {
-					done(err);
+						it('should call a callback with no error', done => {
+							__private
+								.saveBlockStep(
+									{
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 17]
```diff
@@ -330,346 +675,1106 @@ describe('blocks/chain', () => {
...
+					beforeEach(() => {
+						return modules.rounds.tick.callsArgWith(1, 'tick-ERR', null);
+					});
 
-		it('should call modules.transactions.undoUnconfirmedList');
+					it('should call a callback with error', done => {
+						__private
+							.saveBlockStep(
+								{
+									id: 5,
```
Comment (SargeKhan) [inline test/unit/modules/blocks/chain.js]: same

File: test/unit/modules/blocks/chain.js [hunk 18]
```diff
@@ -15,313 +15,596 @@
...
-		describe('applyTransaction', () => {
-			it('should call modules.transactions.applyUnconfirmed');
-
-			it('should call modules.transactions.applyUnconfirmed with transaction');
-
-			it('should call modules.transactions.applyUnconfirmed with sender');
-
-			it('should call modules.transactions.applyUnconfirmed with callback');
+			describe('tx.batch', () => {
+				describe('when fails', () => {
```
Comment (MaciejBaj) [inline test/unit/modules/blocks/chain.js]: Please change here and in the following examples from:
```
describe('tx.batch')
   describe('when fails')
describe('tx.batch')
   describe('when succeeds')
```
to the form which is used in the rest of the tests to stay consistent:
```
describe('when tx.batch fails')
describe('when tx.batch succeeds)
```

File: test/unit/modules/blocks/chain.js [hunk 19]
```diff
@@ -15,313 +15,596 @@
...
+					beforeEach(() => {
+						return modules.accounts.setAccountAndGet.callsArgWith(
+							1,
+							'setAccountAndGet-ERR',
+							true
+						);
+					});
 
-							it('should call self.deleteBlock with oldLastBlock.id');
+					it('should return process.exit(0)', done => {
```
Comment (MaciejBaj) [inline test/unit/modules/blocks/chain.js]: Please change it from `should return process.exit(0)`
to `should call process.exit with 0`
Also, you are not checking it in your tests. Please apply it here and in the following `process.exit(0)` tests.

File: test/unit/modules/blocks/chain.js (hunk 18 above)
Comment (LucasIsasmendi) [reply test/unit/modules/blocks/chain.js]: The pattern is
```
describe('tx.batch')
    describe('when fails)
    describe('when succeeds')
```
It was used in all the test from `modules/blocks` folder, should I change all of them?

Comment (MaciejBaj) [reply test/unit/modules/blocks/chain.js]: There is the only`process.js` file that conforms to this new pattern. `api.js`, `utils.js`, `verify.js` are not using them. Please change the current file only for now, we can change the `process.js` file in another PR.

**Pull request merged**